tools/hls-verifier/lib/HlsVhdlTb.cpp

```diff
@@ -6,664 +6,517 @@
+// - declareSignals: signals between the dual port RAM and the circuit
+// - instantiateRAMModel: instantiate the RAMs that communicate with files
+// - connectToDuv: connect the RAM to the circuit
+//
+// Current assumption: the circuit is always a master device w.r.t. the RAM
+//
+// This struct can be extended / adapted to other interfaces like AXI
+struct MemRefToDualPortRAM {
+  SmallVector<tuple</* Signal name of the circuit */ std::string,
+                    /* Signal name of the DPRAM*/ std::string,
+                    /* Bitwidth */ unsigned>>
+      memrefToDPRAM;
+  mlir::MemRefType type;
+  std::string argName;
+
+  MemRefToDualPortRAM(mlir::MemRefType type, const std::string &argName)
+      : type(type), argName(argName) {
+
+    int dataWidth = type.getElementType().getIntOrFloatBitWidth();
+    int dataDepth = type.getNumElements();
+    int addrWidth = ((int)ceil(log2(dataDepth)));
+
+    // The two_port_RAM has two read/write interfaces, each has
+    // - we: write enable (must be set to 1 when writing)
+    // - ce: clock enable (must be set to 1 for both read and write).
+    // - address: address port
+    // - din: store data
+    // - dout: read data
+    // Currently, our memory controler assumes that there is only one read
+    // channel and one write channel. Therefore, many signals are not used
+
+    // This mapping only records the used signals.
+    memrefToDPRAM.emplace_back("storeEn", WE0_PORT, 1);
+    memrefToDPRAM.emplace_back("storeData", D_IN0_PORT, dataWidth);
+    memrefToDPRAM.emplace_back("storeAddr", ADDR0_PORT, addrWidth);
+    memrefToDPRAM.emplace_back("loadEn", CE1_PORT, 1);
+    memrefToDPRAM.emplace_back("loadData", D_OUT1_PORT, dataWidth);
+    memrefToDPRAM.emplace_back("loadAddr", ADDR1_PORT, addrWidth);
+  }
+
+  void declareConstants(mlir::raw_indented_ostream &os,
+                        const std::string &inputVectorPath,
+                        const std::string &outputFilePath) {
+    int dataWidth = type.getElementTypeBitWidth();
+    int dataDepth = type.getNumElements();
+    int addrWidth = ((int)ceil(log2(dataDepth)));
+    declareConstant(os, "INPUT_" + argName, "STRING",
+                    "\"" + inputVectorPath + "/input_" + argName + ".dat" +
+                        "\"");
+    declareConstant(os, "OUTPUT_" + argName, "STRING",
+                    "\"" + outputFilePath + "/output_" + argName + ".dat" +
+                        "\"");
+    declareConstant(os, "DATA_WIDTH_" + argName, "INTEGER",
+                    to_string(dataWidth));
+    declareConstant(os, "ADDR_WIDTH_" + argName, "INTEGER",
+                    to_string(addrWidth));
+    declareConstant(os, "DATA_DEPTH_" + argName, "INTEGER",
+                    to_string(dataDepth));
+  }
+
+  // Declare signals appear in the circuit interface
+  void declareSignals(mlir::raw_indented_ostream &os) {
+    for (auto [_, sigName, bitwidth] : memrefToDPRAM) {
```

Comment:
My clang-tidy suggests

```suggestion
    for (const auto &[_, sigName, bitwidth] : memrefToDPRAM) {
```

due to the performance issue.
The same warning appears in almost all similar for loops